daemon/logger/copier_test.go

```diff
@@ -205,6 +223,38 @@ func TestCopierSlow(t *testing.T) {
 	}
 }
 
+func TestCopierWithSized(t *testing.T) {
+	var jsonBuf bytes.Buffer
+	sizedLogger := &TestSizedLoggerJSON{Encoder: json.NewEncoder(&jsonBuf)}
+
+	logbuf := bytes.NewBufferString(strings.Repeat(".", sizedLogger.BufSize()*2))
+	rd, wr := io.Pipe()
+	c := NewCopier(map[string]io.Reader{"stdout": rd}, sizedLogger)
+	c.Run()
+	go func() {
+		io.Copy(wr, logbuf)
+		c.Wait()
```

Comment:
We need this outside of the goroutine, I think?
Otherwise it will (potentially) decode from the buffer and get an EOF before it's really done.

Reply:
We may need a Wait both in the goroutine and outside of it to ensure the Copier doesn't close early while reading the data from the pipe. I think there's the possibility that during the read, given the right size and read offsets, that the Copier could check the closed status of the Copier and cease reading from the pipe input (shouldn't be tickled in this test case but is a point to note). I will add the Wait outside the goroutine as well and see if I can easily cause the test to fail.

Reply:
Alternatively we could ditch the buffer and read directly from the pipe below.

Reply:
I'm not sure I follow you on that, can you elaborate some more?

Reply:
Sorry, forget about that. I brain-farted on what we were doing here.

Reply:
I believe I've found a more robust and simpler approach. I've pushed that change now. @cpuguy83 PTAL.

Reply:
Oh yes, much simpler.